crates/sui-node/src/metrics.rs

```diff
@@ -34,3 +37,32 @@ async fn metrics(Extension(registry): Extension<Registry>) -> (StatusCode, Strin
         ),
     }
 }
+
+#[derive(Clone)]
+pub struct GrpcMetrics {
+    inflight_grpc: IntGaugeVec,
+}
+
+impl GrpcMetrics {
+    pub fn new(registry: &Registry) -> Self {
+        Self {
+            inflight_grpc: register_int_gauge_vec_with_registry!(
+                "inflight_grpc",
+                "Total in-flight GRPC per route",
+                &["path"],
+                registry,
+            )
+            .unwrap(),
+        }
+    }
+}
+
+impl MetricsCallbackProvider for GrpcMetrics {
+    fn on_request(&self, path: String) {
+        self.inflight_grpc.with_label_values(&[&path]).inc();
+    }
+
+    fn on_response(&self, path: String, _latency: Duration, _status: u16, _grpc_status_code: Code) {
+        self.inflight_grpc.with_label_values(&[&path]).dec();
```

Comment:
I don't think this will run if the client disconnects before receiving the response, which will cause these counters to gradually inflate.

Reply:
ah that's a good point, we'll probably need to use an RAII here to handle that case. We could employ similar logic to how the global concurrency limit layer works: https://docs.rs/tower/latest/src/tower/limit/concurrency/layer.rs.html#38-40

Reply:
if we go with that route we probably could just write our own tower layer to better keep track of these things